kafka-handler/src/java/org/apache/hadoop/hive/kafka/KafkaDagCredentialSupplier.java

```diff
@@ -0,0 +1,136 @@
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaDagCredentialSupplier.class);
+  private static final Text KAFKA_DELEGATION_TOKEN_KEY = new Text("KAFKA_DELEGATION_TOKEN");
+
+  @Override
+  public Token<?> obtainToken(BaseWork work, Set<TableDesc> fileSinkTableDescs, Configuration conf) {
+    if(!(work instanceof MapWork)){
+      return null;
+    }
+    Map<String, PartitionDesc> partitions = ((MapWork) work).getAliasToPartnInfo();
+
+    // We don't need to iterate on all partitions, and check the same TableDesc.
+    PartitionDesc partition = partitions.values().stream().findFirst().orElse(null);
+    if (partition != null) {
+      TableDesc tableDesc = partition.getTableDesc();
+      if (collectKafkaDelegationTokenForTableDesc(tableDesc)) {
+        // don't collect delegation token again, if it was already successful
+        return getKafkaDelegationTokenForBrokers(conf, tableDesc);
+      }
+    }
+
+    for (TableDesc tableDesc : fileSinkTableDescs) {
+      if (collectKafkaDelegationTokenForTableDesc(tableDesc)) {
+        // don't collect delegation token again, if it was already successful
+        return getKafkaDelegationTokenForBrokers(conf, tableDesc);
+      }
+    }
+    return null;
+  }
+
+  @Override
+  public Text getTokenAlias() {
+    return KAFKA_DELEGATION_TOKEN_KEY;
+  }
+
+  /**
+   * Returns whether we should collect delegation tokens for kafka in the scope of a TableDesc.
+   * If "security.protocol" is set to "PLAINTEXT", we don't need to collect delegation token at all.
+   * @param tableDesc
+   * @return true if we should collect a token for the specified table and false otherwise.
+   */
+  private boolean collectKafkaDelegationTokenForTableDesc(TableDesc tableDesc) {
+    String kafkaBrokers = (String) tableDesc.getProperties().get("kafka.bootstrap.servers"); //FIXME: KafkaTableProperties
+    String consumerSecurityProtocol = (String) tableDesc.getProperties().get(
+        "kafka.consumer." + CommonClientConfigs.SECURITY_PROTOCOL_CONFIG);
+    String producerSecurityProtocol = (String) tableDesc.getProperties().get(
+        "kafka.producer." + CommonClientConfigs.SECURITY_PROTOCOL_CONFIG);
+    return kafkaBrokers != null && !kafkaBrokers.isEmpty()
+        && !CommonClientConfigs.DEFAULT_SECURITY_PROTOCOL.equalsIgnoreCase(consumerSecurityProtocol)
+        && !CommonClientConfigs.DEFAULT_SECURITY_PROTOCOL.equalsIgnoreCase(producerSecurityProtocol);
+  }
+
+  private Token<?> getKafkaDelegationTokenForBrokers(Configuration conf, TableDesc tableDesc) {
+    String kafkaBrokers = (String) tableDesc.getProperties().get("kafka.bootstrap.servers");
+    LOG.info("Getting kafka credentials for brokers: {}", kafkaBrokers);
+
+    String keytab = HiveConf.getVar(conf, HiveConf.ConfVars.HIVE_SERVER2_KERBEROS_KEYTAB);
+    String principal = HiveConf.getVar(conf, HiveConf.ConfVars.HIVE_SERVER2_KERBEROS_PRINCIPAL);
+    try {
+      principal = SecurityUtil.getServerPrincipal(principal, "0.0.0.0");
```

Comment:
Can we call getLocalHostName() directly here instead of sending 0.0.0.0, we already have the conf obj which is required by the method.

Reply:
I suppose you mean to refactor the call using `org.apache.hadoop.security.SecurityUtil#getLocalHostName` but this method is not public. The method is package private so it cannot be called directly.

Moreover, the `SecurityUtil.getServerPrincipal(principal, "0.0.0.0")` is widely used in our code base so not sure why we should not use it here.

Reply:
also, SecurityUtil.getServerPrincipal(principal, "0.0.0.0") is introduced by an earlier patch, not in this refactor, I believe if it's not okay (but it is), we need to address it in another ticket instead of hiding it in a refactor